### ISSUE TITLE
Fix random test faulires

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1478]
There are several tests that fail randomly, usually only on the Travis CI infrastrucrure. Let's fix them, or replace them with others so that they don't do more harm than good.

- [ ] Path: Path pattern basic tests
- [ ] User admin test case
- [ ] Language settings file not valid JSON

**Comment by quicksketch** [comment]
I got a couple more randoms today. I think some of these new ones might be related to Background Fetch if the server isn't running fast enough. This was on PHP 5.4 and the tests ran for 30 minutes, compared to the normally speedy 10-15 minutes.

```

---- Bootstrap: Boot and exit hook invocation (HookBootExitTestCase) ----


Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      bootstrap.test     377 HookBootExitTestCase->testHookBootE
    Page cache cleared.
Fail      Other      bootstrap.test     380 HookBootExitTestCase->testHookBootE
    hook_boot called with aggressive cache and no cached page.
Fail      Other      bootstrap.test     381 HookBootExitTestCase->testHookBootE
    hook_exit called with aggressive cache and no cached page.


---- Layout: Layout Interface Test (LayoutInterfaceTest) ----


Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      layout.test        608 LayoutInterfaceTest->testMainConten
    "The &quot;Main page content&quot; block may only be added once to this
    layout." found
Exception Notice     layout.test        614 LayoutInterfaceTest->testMainConten
    Undefined offset: 0
Exception Notice     layout.test        618 LayoutInterfaceTest->testMainConten
    Undefined offset: 0
Fail      Other      layout.test        621 LayoutInterfaceTest->testMainConten
    Found the Save layout button
Fail      Other      layout.test        621 LayoutInterfaceTest->testMainConten
    Found the requested form fields at
Fail      Other      layout.test        622 LayoutInterfaceTest->testMainConten
    Layout saved successfully after removing old content block.
Fail      Other      layout.test        632 LayoutInterfaceTest->testMainConten
    Main content shown in sidebar after moving main content block.


---- Module: Module API (ModuleUnitTest) ----


Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      module.test         84 ModuleUnitTest->testModuleImplement
    The module implements cache is populated after requesting a page.
```

**Comment by quicksketch** [comment]
Details about the "User admin test case" fail:

```
---- User: User administration (UserAdminTestCase) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      user.test         1441 UserAdminTestCase->testUserAdmin() 
    User C blocked
Fail      E-mail     user.test         1450 UserAdminTestCase->testUserAdmin() 
    Activation mail sent to user C
```

**Comment by quicksketch** [comment]
Filed a PR at https://github.com/backdrop/backdrop/pull/1200 that addresses 4 of these failures. Updated the summary.

**Comment by quicksketch** [comment]
I think I've got all these worked out. For image styles and module API, the added sleep() timeouts should help the slower servers successfully pass when dealing with time-sensitive issues. The user and path alias tests I rewrote to use more robust XPath checking of elements instead of guessing based on the User ID.

The `Layout: Layout Interface Test (LayoutInterfaceTest)` is the only one on which I'm unclear. If that test fail keeps coming up, hopefully we will have at least narrowed the scope of the random test failures.

All the fixes so far have gone into 1.2.x and 1.x, so I'm tagging this for 1.2.4.

**Comment by quicksketch** [comment]
Here's dump of the current fail on the Layout Interface Test:

```
---- Layout: Layout Interface Test (LayoutInterfaceTest) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      layout.test        608 LayoutInterfaceTest->testMainConten
    "The &quot;Main page content&quot; block may only be added once to this
    layout." found
Exception Notice     layout.test        614 LayoutInterfaceTest->testMainConten
    Undefined offset: 0
Exception Notice     layout.test        618 LayoutInterfaceTest->testMainConten
    Undefined offset: 0
Fail      Other      layout.test        621 LayoutInterfaceTest->testMainConten
    Found the Save layout button
Fail      Other      layout.test        621 LayoutInterfaceTest->testMainConten
    Found the requested form fields at
Fail      Other      layout.test        622 LayoutInterfaceTest->testMainConten
    Layout saved successfully after removing old content block.
Fail      Other      layout.test        632 LayoutInterfaceTest->testMainConten
    Main content shown in sidebar after moving main content block.
```

Looks like our old friend "Invalid JSON" is still happening as well:
```
---- User: User email administration (UserAdminSettingsEmailFormTest) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Exception Uncaught e config.inc         898 ConfigFileStorage->read()          
    ConfigStorageReadException: The configuration file &quot;user.mail&quot; is
    not properly formatted JSON. in ConfigFileStorage-&gt;read() (line 898 of
    /home/travis/build/jenlampton/backdrop/core/includes/config.inc).

**Comment by quicksketch** [comment]
We're still getting a lot of failing page cache tests:

```
---- Bootstrap: Page cache test (BootstrapPageCacheTestCase) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------

Fail      Other      bootstrap.test     204 BootstrapPageCacheTestCase->testPag
    Cached page request.
````

I think this might have to do with the background fetch functionality and our use of `sleep()` to try and wait long enough, but that is not guaranteed, especially with the slower Travis CI servers. I filed a new PR at https://github.com/backdrop/backdrop/issues/1367.

**Comment by quicksketch** [comment]
I merged in https://github.com/backdrop/backdrop/pull/1367, which passed the fist attempt. We'll see how it does with new PRs as they come in.

**Comment by Gormartsen** [comment]
I am able to reproduce next issue with backdrop/backdrop#1367  applied.
```
Bootstrap: Page cache test (BootstrapPageCacheTestCase) 155 passes, 1 fail, 0 exceptions, and 41 debug messages

Test run duration: 58 sec

Detailed test results
---------------------


---- Bootstrap: Page cache test (BootstrapPageCacheTestCase) ----


Status    Group      Filename          Line Function
--------------------------------------------------------------------------------
Fail      Other      bootstrap.test     202 BootstrapPageCacheTestCase->testPag
    Initial page requests returned before shutdown functions are executed.
```

I am doing investigation here.

**Comment by Gormartsen** [comment]
PS: random fail happen again here: https://travis-ci.org/backdrop/backdrop/jobs/130238636

**Comment by Gormartsen** [comment]
This test is failing here:
```

**Comment by quicksketch** [comment]
> It mean that [...] takes 5.03 sec to process

Well in this case it's a "good" failure! Because the functionality is actually broken or not working. The page isn't being returned immediately, it's getting held up by a shutdown function when it shouldn't be. We've had some issues where `backdrop_page_footer()` doesn't flush and close the connection properly based on differing PHP configurations (e.g. mod_php vs. FastCGI vs. PHP-FPM). So this likely needs to be fixed in `backdrop_page_footer()` rather than adjusting the test.

**Comment by Gormartsen** [comment]
Thank you for details @quicksketch . It is different setup from travis. I am going to figure it out this one.

**Comment by Gormartsen** [comment]
@quicksketch I fixed this issue. PR is here: https://github.com/backdrop/backdrop/pull/1378

**Comment by quicksketch** [comment]
It looks like my recent fixes to the page cache (adding `sleep()` and disabling the background fetch) missed the `testPageCompression()` test, which is failing regularly in both PHP 5.3 and 7.0.

```
---- Bootstrap: Page cache test (BootstrapPageCacheTestCase) ----

Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      bootstrap.test     328 BootstrapPageCacheTestCase->testPag
    Page was cached.
```
I've filed another PR at https://github.com/backdrop/backdrop/pull/1383. Perhaps combined with yours, we'll get the page cache tests fixed entirely.

**Comment by quicksketch** [comment]
> Perhaps combined with yours, we'll get the page cache tests fixed entirely.

We haven't had any page cache failures since this change. Yay! Fewer failures. We are still getting several regularly though, let's keep at it.

**Comment by quicksketch** [comment]
PHP 7 gets this error fairly frequently:

```
---- SimpleTest: SimpleTest functionality (SimpleTestFunctionalTest) ----

Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Exception Notice     simpletest.module  219 _simpletest_batch_finished()       
    Undefined offset: 0
```

And PHP 5.3 gets this one fairly regularly still:
```
---- User: User email administration (UserAdminSettingsEmailFormTest) ----

Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Exception Uncaught e config.inc         898 ConfigFileStorage->read()          
    ConfigStorageReadException: The configuration file &quot;user.mail&quot; is
    not properly formatted JSON. in ConfigFileStorage-&gt;read() (line 898 of
    /home/travis/build/backdrop/backdrop/core/includes/config.inc).
```

**Comment by klonos** [comment]
I constantly seem to be getting this in the php70 tests of my PRs:

Layout: Layout BlockText Test (LayoutBlockTextTest)

- `exception` Undefined index: file **layout.test**:1723
- `fail` File usage recorded for the file within the block. **layout.test**:1723
- `fail` File has been marked permanent by its file usage. **layout.test**:1724
- `exception` Undefined index: file **layout.test**:1743
- `fail` File usage is unaffected by removing the block from the layout. **layout.test**:1743
- `fail` File still exists and is marked as permanent after removing the block from the layout. **layout.test**:1744

```
---- Layout: Layout BlockText Test (LayoutBlockTextTest) ---- Status Group Filename Line Function -------------------------------------------------------------------------------- Exception Notice     layout.test       1723 LayoutBlockTextTest->testBlockText(
Undefined index: file Fail      Other      layout.test       1723 LayoutBlockTextTest->testBlockText(
File usage recorded for the file within the block. Fail      Other      layout.test       1724 LayoutBlockTextTest->testBlockText(
File has been marked permanent by its file usage. Exception Notice     layout.test       1743 LayoutBlockTextTest->testBlockText(
Undefined index: file Fail      Other      layout.test       1743 LayoutBlockTextTest->testBlockText(
File usage is unaffected by removing the block from the layout. Fail      Other      layout.test       1744 LayoutBlockTextTest->testBlockText(
File still exists and is marked as permanent after removing the block from the layout.
```

**Comment by klonos** [comment]
@quicksketch tiny coding standards fix, follow-up to your PR: https://github.com/backdrop/backdrop/pull/1945

**Comment by quicksketch** [comment]
Thanks @klonos! Merged that space fix to 1.x and 1.7.x :+1:

**Comment by klonos** [comment]
@quicksketch I am still getting very frequent `fail comment-new class not found. comment.test:628` failures. Do you think perhaps worth increasing the time in https://github.com/backdrop/backdrop/blob/1.x/core/modules/comment/tests/comment.test#L575 ?

**Comment by klonos** [comment]
Increase from 2 to 3 seconds: https://github.com/backdrop/backdrop/pull/2357

**Comment by quicksketch** [comment]
> I am still getting very frequent fail comment-new class not found. comment.test:628 failures.

I encountered this again today. It's worth a shot to increase the timeout I guess, though I don't know if it will have any impact. Merged https://github.com/backdrop/backdrop/pull/2357 into 1.x and 1.11.x.

**Comment by jenlampton** [comment]
I still sometimes get a random failure with `BatchProcessingTestCase`.

**Comment by jenlampton** [comment]
Problems with CommentInterfaceTest: 

```
---- Comment: Comment interface (CommentInterfaceTest) ----


Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      comment.test       628 CommentInterfaceTest->testCommentCl
    comment-new class not found.
```

**Comment by quicksketch** [comment]
New PR for fixing the `comment-new` race-condition that causes what might be our most common random failure: https://github.com/backdrop/backdrop/pull/2601

**Comment by herbdool** [comment]
RTBC for this PR https://github.com/backdrop/backdrop/pull/2601#issuecomment-485200847 @quicksketch

**Comment by klonos** [comment]
From time to time, I see this:

`"Generated 20 URL aliases." found path_pattern.test:929`

**Comment by klonos** [comment]
...never mind; was already in the issue summary as `Path: Path pattern bulk updating (PathPatternBulkUpdateTestCase)`

**Comment by quicksketch** [comment]
A new PR for a failure that's been happening frequently since 1.16.0: https://github.com/backdrop/backdrop/pull/3172

**Comment by indigoxela** [comment]
Seems like CommentInterfaceTest is back...

**Comment by quicksketch** [comment]
We've made a lot of progress recently since the switch to GitHub Actions (a lot of these had been showing up more frequently). The following now seem resolved:


- Random test failures with PathPatternBulkUpdateTestCase #4862
- Random failures or timeouts in BatchProcessingTestCase (race condition in all batch processing) #5205
- Random failures in CommentInterfaceTest - Link with label 1 new comment not found. #5189 

The fixing of #5205 should fix intermittent failures in several other tests (anything that used batch processing). I think this may include #1965, but I'll leave that issue open to confirm.

**Comment by quicksketch** [comment]
According to the issue summary, these failures are still giving us trouble:

>  AJAX: AJAX forms on cached pages
> Form API: Forms using form storage on cached page
> SimpleTest: Installation profile module tests
> Image: Image style flood protection (ImageStyleFloodProtection):
> `fail Correct number of images (1) return 403 access denied. image.test:1792`

But I haven't seen any of this test failures in recent memory. After seemingly fixing https://github.com/backdrop/backdrop-issues/issues/7086, we may not have any random failures in our test coverage; wouldn't that be something?

@backdrop/core-committers Let's keep an eye out for any remaining random test failures. If we don't see any in the next two weeks, I think this overall "meta" issue can be closed. Any new issues that occur we can handle one at a time without the meta issue. This issue has been open for 10 years.

**Comment by indigoxela** [comment]
I just rebased several PRs and none of them had any random test failure. I belief, it's finally time to close this issue.

The last random buddies I'm aware of were the *UpgradePathTestCase ones and those are fixed now.

Didn't see any of those remaining ones from the issue description recently, so I'm assuming those all got fixed in the meantime.

**Comment by quicksketch** [comment]
Closing as completed 🎉 

For new random test failures, please open individual issues. If you feel like cross-linking this issue or adding it as a parent issue, please continue to do so, but leave this issue closed.
